runtime/tests/interpreter/values_test.go

```diff
@@ -0,0 +1,1547 @@
+package interpreter
+
+import (
+	"flag"
+	"fmt"
+	"math"
+	"math/big"
+	"math/rand"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/onflow/atree"
+
+	"github.com/onflow/cadence/runtime/ast"
+	"github.com/onflow/cadence/runtime/common"
+	"github.com/onflow/cadence/runtime/interpreter"
+	"github.com/onflow/cadence/runtime/sema"
+	"github.com/onflow/cadence/runtime/tests/utils"
+)
+
+// TODO: make these program args?
+const containerMaxDepth = 3
+const containerMaxSize = 1_000
+const innerContainerMaxSize = 300
+const compositeMaxFields = 10
+
+var runSmokeTests = flag.Bool("runSmokeTests", true, "Run smoke tests on values")
+
+func TestRandomMapOperations(t *testing.T) {
+	if !*runSmokeTests {
+		t.SkipNow()
+	}
+
+	seed := time.Now().UnixNano()
+	fmt.Printf("Seed used for map opearations test: %d \n", seed)
+	rand.Seed(seed)
+
+	storage := interpreter.NewInMemoryStorage()
+	inter, err := interpreter.NewInterpreter(
+		&interpreter.Program{
+			Program:     ast.NewProgram([]ast.Declaration{}),
+			Elaboration: sema.NewElaboration(),
+		},
+		utils.TestLocation,
+		interpreter.WithStorage(storage),
+		interpreter.WithImportLocationHandler(
+			func(inter *interpreter.Interpreter, location common.Location) interpreter.Import {
+				return interpreter.VirtualImport{
+					Elaboration: inter.Program.Elaboration,
+				}
+			},
+		),
+	)
+	require.NoError(t, err)
+
+	numberOfValues := randomInt(containerMaxSize)
+
+	var testMap, copyOfTestMap *interpreter.DictionaryValue
+	var storageSize, slabCounts int
+
+	entries := newValueMap(numberOfValues)
+	orgOwner := common.Address{'A'}
+
+	t.Run("construction", func(t *testing.T) {
+		keyValues := make([]interpreter.Value, numberOfValues*2)
+		for i := 0; i < numberOfValues; i++ {
+			key := randomHashableValue(inter)
+			value := randomStorableValue(inter, 0)
+
+			entries.put(inter, key, value)
+
+			keyValues[i*2] = key
+			keyValues[i*2+1] = value
+		}
+
+		testMap = interpreter.NewDictionaryValueWithAddress(
+			inter,
+			interpreter.DictionaryStaticType{
+				KeyType:   interpreter.PrimitiveStaticTypeAnyStruct,
+				ValueType: interpreter.PrimitiveStaticTypeAnyStruct,
+			},
+			orgOwner,
+			keyValues...,
+		)
+
+		storageSize, slabCounts = getSlabStorageSize(t, storage)
+
+		require.Equal(t, testMap.Count(), entries.size())
+
+		entries.foreach(func(orgKey, orgValue interpreter.Value) (exit bool) {
+			exists := testMap.ContainsKey(inter, interpreter.ReturnEmptyLocationRange, orgKey)
+			require.True(t, bool(exists))
+
+			value, found := testMap.Get(inter, interpreter.ReturnEmptyLocationRange, orgKey)
+			require.True(t, found)
+			utils.AssertValuesEqual(t, inter, orgValue, value)
+
+			return false
+		})
+
+		owner := testMap.GetOwner()
+		assert.Equal(t, orgOwner[:], owner[:])
```

Comment:
Is it possible to compare the values directly or is the slicing necessary?

Reply:
Yes, should be possible to compare directly